agent/lib/kontena/rpc_client.rb

```diff
@@ -0,0 +1,75 @@
+require_relative 'logging'
+
+module Kontena
+  class RpcClient
+    include Celluloid
+    include Kontena::Logging
+
+    class Error < StandardError
+      attr_accessor :code, :message, :backtrace
+
+      def initialize(code, message, backtrace = nil)
+        self.code = code
+        self.message = message
+        self.backtrace = backtrace
+      end
+    end
+
+    class TimeoutError < Error; end
+
+    attr_reader :requests
+
+    # @param [Kontena::WebsocketClient] client
+    def initialize(client)
+      @requests = {}
+      @client = client
+      info 'initialized'
+    end
+
+    # @param [String] method
+    # @param [Array] params
+    def notification(method, params)
+      @client.send_notification(method, params)
+    rescue => exc
+      logger.error exc.message
+    end
+
+    # @param [String] method
+    # @param [Array] params
+    # @return [Celluloid::Future]
+    def request(method, params)
+      id = request_id
+      @requests[id] = nil
+      future = Celluloid::Future.new {
+        sleep 0.01 until @client.connected?
+        @client.send_request(id, method, params)
+        time = Time.now.utc
+        until !@requests[id].nil?
+          sleep 0.01
+          raise TimeoutError.new(500, 'Request timed out') if time < (Time.now.utc - 30)
```

Comment:
This leaks the `@requests[id]`... but it has to, if it wants to avoid re-using that `request_id` :)